### DISCUSSION TITLE
Data loading with padding of an average number of nodes

### DISCUSSION THREAD

**Discussion by shrimonmuke0202** [pyg-team/pytorch_geometric#2995] [Q&A]
How can I fix the padding of edge_index in the data loader for a fixed dimension?
For e.g:  if my maximum number of nodes in the graph is 46 but I want all the edge indexes with an average number of nodes which is around 43. In this case, how can I fix the padding length of each batch with a fixed size 43?

**Comment by rusty1s** [comment]
Can you give me an example on what you are trying to do exactly? If you want to have all graphs to be of size `43`, you need to either pad/filter out features in `x` and edges in `edge_index` individually:
```python
def fix_dimension(data):
    if data.num_nodes < 43:
        data.x = torch.cat([data.x, torch.zeros(43 - data.num_nodes, data.x.size(-1))], dim=0)
        mask = (data.edge_index[0] < 43) & (data.edge_index[1] < 43)
        data.edge_index = data.edge_index[:, mask]
    elif data.num_nodes > 43:
       data.x = data.x[:43]
    return data

dataset = MyDataset(transform=fix_dimension)
```

**Reply by shrimonmuke0202** [reply]
I want all graphs with edge_index and features with the same number of nodes.

class TestbedDataset(InMemoryDataset):
    def __init__(self, root='/tmp', dataset='davis', 
                 xd=None, xt=None, y=None, transform=None,
                 pre_transform=None,smile_graph=None):

        #root is required for save preprocessed data, default is '/tmp'
        super(TestbedDataset, self).__init__(root, transform, pre_transform)
        # benchmark dataset, default = 'davis'
        self.dataset = dataset
        if os.path.isfile(self.processed_paths[0]):
            print('Pre-processed data found: {}, loading ...'.format(self.processed_paths[0]))
            self.data, self.slices = torch.load(self.processed_paths[0])
        else:
            print('Pre-processed data {} not found, doing pre-processing...'.format(self.processed_paths[0]))
            self.process(xd, xt, y,smile_graph)
            self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        pass
        #return ['some_file_1', 'some_file_2', ...]

    @property
    def processed_file_names(self):
        return [self.dataset + '.pt']

    def download(self):
        # Download to `self.raw_dir`.
        pass

    def _download(self):
        pass

    def _process(self):
        if not os.path.exists(self.processed_dir):
            os.makedirs(self.processed_dir)

    # Customize the process method to fit the task of drug-target affinity prediction
    # Inputs:
    # XD - list of SMILES, XT: list of encoded target (categorical or one-hot),
    # Y: list of labels (i.e. affinity)
    # Return: PyTorch-Geometric format processed data
    def process(self, xd, xt, y,smile_graph):
        assert (len(xd) == len(xt) and len(xt) == len(y)), "The three lists must be the same length!"
        data_list = []
        data_len = len(xd)
        for i in range(data_len):
            print('Converting SMILES to graph: {}/{}'.format(i+1, data_len))
            smiles = xd[i]
            target = xt[i]
            labels = y[i]
            # convert SMILES to molecular representation using rdkit
            c_size,features, edge_index = smile_graph[smiles]
            
            # make the graph ready for PyTorch Geometrics GCN algorithms:
            GCNData = DATA.Data(x=torch.Tensor(features),
                                edge_index=torch.LongTensor(edge_index).transpose(1, 0),
                                y=torch.FloatTensor([labels]))
            GCNData.target = torch.LongTensor([target])
            GCNData.__setitem__('c_size', torch.LongTensor([c_size]))
            # append graph, label and target sequence to data list
            data_list.append(GCNData)

        if self.pre_filter is not None:
            data_list = [data for data in data_list if self.pre_filter(data)]

        if self.pre_transform is not None:
            data_list = [self.pre_transform(data) for data in data_list]
        print('Graph construction done. Saving to file.')
        data, slices = self.collate(data_list)
        # save preprocessed data:
        torch.save((data, slices), self.processed_paths[0])

This is my own dataset. Where I will apply the function fix dimension?

**Reply by rusty1s** [reply]
You can simply pass it as the `pre_transform` or `transform` argument.

**Reply by shrimonmuke0202** [reply]
Using the fix dimension function can I get all graphs of a same number of nodes in every batch?

**Reply by rusty1s** [reply]
Yes, this will make all graphs be of fixed size (in this case 43).

**Reply by shrimonmuke0202** [reply]
I want the shape of all  x and adj must be the same in all batch

**Reply by rusty1s** [reply]
I see. Doing
```python
def fix_dimension(data):
    if data.num_nodes < 28:
        data.x = torch.cat([data.x, torch.zeros(28 - data.num_nodes, data.x.size(-1))], dim=0)
        mask = (data.edge_index[0] < 28) & (data.edge_index[1] < 28)
        data.edge_index = data.edge_index[:, mask]
    elif data.num_nodes > 28:
       data.x = data.x[:28]
    data.num_nodes = 28
    return data
```
should fix this.

**Reply by shrimonmuke0202** [reply]
If I take 22 since the minimum number of nodes is 22 in the graph then the code is not executed in the google colab i.e session crash occurs,

why this happens

**Reply by shrimonmuke0202** [reply]
The session does not crash for 26 but for 22 the session crash occurs

**Reply by rusty1s** [reply]
Oh yeah, I should not write code without testing it :D The correct code is
```python
import torch
from torch_geometric.data import DataLoader
from torch_geometric.datasets import TUDataset
from torch_geometric.utils import to_dense_batch, to_dense_adj

N = 22


def fix_dimension(data):
    if data.num_nodes < N:
        data.x = torch.cat(
            [data.x, torch.zeros(N - data.num_nodes, data.x.size(-1))], dim=0)
    elif data.num_nodes > N:
        data.x = data.x[:N]
        mask = (data.edge_index[0] < N) & (data.edge_index[1] < N)
        data.edge_index = data.edge_index[:, mask]
    data.num_nodes = N
    print(data.edge_index.max())
    return data


dataset = TUDataset('/tmp/MUTAG', name='MUTAG', transform=fix_dimension)
loader = DataLoader(dataset, batch_size=32, drop_last=True)

for data in loader:
    x, _ = to_dense_batch(data.x, data.batch)
    adj = to_dense_adj(data.edge_index, data.batch)
    print(x.shape, adj.shape)
```

**Comment by shrimonmuke0202** [comment]
Thank you this code works.
